x-pack/apm-server/sampling/eventstorage/storage_manager.go

```diff
@@ -107,16 +115,121 @@ func (s *StorageManager) runValueLogGC(discardRatio float64) error {
+// If so, it drops and recreates the underlying badger DB.
+// The loop stops when it receives from stopping.
+func (s *StorageManager) RunDropLoop(stopping <-chan struct{}, ttl time.Duration, storageLimitInBytes uint64) error {
+	select {
+	case <-stopping:
+		return nil
+	case s.dropLoopCh <- struct{}{}:
+	}
+	defer func() {
+		<-s.dropLoopCh
+	}()
+
+	if storageLimitInBytes == 0 {
+		<-stopping
+		return nil
+	}
+
+	timer := time.NewTicker(min(time.Minute, ttl)) // Eval db size every minute as badger reports them with 1m lag, but use min to facilitate testing
+	defer timer.Stop()
+	var firstExceeded time.Time
+	for {
+		select {
+		case <-stopping:
+			return nil
+		case <-timer.C:
+			lsm, vlog := s.Size()
+			if uint64(lsm+vlog) >= storageLimitInBytes { //FIXME: Add a bit of buffer? Is s.storage.pendingSize reliable enough?
+				now := time.Now()
+				if firstExceeded.IsZero() {
+					firstExceeded = now
+					s.logger.Warnf("badger db size has exceeded storage limit; db will be dropped and recreated if problem persists for `sampling.tail.ttl` (%s)", ttl.String())
+				}
+				if now.Sub(firstExceeded) >= ttl {
+					s.logger.Warnf("badger db size has exceeded storage limit for over `sampling.tail.ttl` (%s), please consider increasing `sampling.tail.storage_limit`; dropping and recreating badger db to recover", ttl.String())
+					err := s.dropAndRecreate()
+					if err != nil {
+						return fmt.Errorf("error dropping and recreating badger db to recover storage space: %w", err)
+					}
+					s.logger.Info("badger db dropped and recreated")
+				}
+			} else {
+				firstExceeded = time.Time{}
+			}
+		}
+	}
+}
+
 func (s *StorageManager) Close() error {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
 	s.rw.Close()
 	return s.db.Close()
 }
 
 // Size returns the db size
+//
+// Caller should either be main Run loop or should be holding RLock already
 func (s *StorageManager) Size() (lsm, vlog int64) {
 	return s.db.Size()
 }
 
+func getBackupPath(path string) string {
+	return filepath.Join(filepath.Dir(path), filepath.Base(path)+".old")
+}
+
+// dropAndRecreate deletes the underlying badger DB at a file system level, and replaces it with a new badger DB.
+func (s *StorageManager) dropAndRecreate() error {
+	backupPath := getBackupPath(s.storageDir)
+	if err := os.RemoveAll(backupPath); err != nil && !errors.Is(err, os.ErrNotExist) {
+		return fmt.Errorf("error removing existing backup dir: %w", err)
+	}
+
+	// FIXME: defer lock release as errors are now returned
+	s.mu.Lock()
+	s.rw.Close()
+	err := s.db.Close()
+	if err != nil {
+		return fmt.Errorf("error closing badger db: %w", err)
+	}
+
+	s.subscriberPosMu.Lock()
+	err = os.Rename(s.storageDir, backupPath)
+	if err != nil {
+		return fmt.Errorf("error backing up existing badger db: %w", err)
+	}
+
+	// Since subscriber position file lives in the same tail sampling directory as badger DB,
+	// Create tail sampling dir, move back subscriber position file, as it is not a part of the DB.
```

Comment:
Could we change the subcriber position directory to be a separate directory than badger data? All the orchestration required to move this file is just not needed.

Reply:
We'll still need to maintain backwards compatibility. e.g. even if we ship a migration in 8.16.3, we'll need to keep backwards compatibility code at apm-server startup time to read from a pre-8.16.3 directory structure and move to a 8.16.3 directory structure. Then the dropAndRecreate implementation can be much cleaner. I was originally thinking about doing the move in 9.0.

Reply:
> even if we ship a migration in 8.16.3, we'll need to keep backwards compatibility code at apm-server startup time to read from a pre-8.16.3 directory structure and move to a 8.16.3 directory structure

Hmm, is it critical for us to keep the subscriber position data for restarts? Will it result in starting from zero and rereading all sampled traces?

Reply:
> Will it result in starting from zero and rereading all sampled traces?

Not 100% sure. Need to dig deeper.

> is it critical for us to keep the subscriber position data for restarts?

Either a performance hit to reread all sampling decisions, or data loss if it does not reread sampling decisions made during apm-server restart.

Reply:
Will this be a blocker? I think that the scope of this PR is fairly large already, not sure if we should further expand the scope.

Reply:
Definitely not a blocker though I don't personally like either of the renames (this one + the one we do earlier to the directory). But, feel free to go ahead with merge if you have approving reviews as I might not be able to get to this today.

Reply:
The reason I'm raising this is that in an ideal world instead of subscriber position being in a subdirectory, I think we should move badger into a subdirectory. And the risk factor may be different.

Reply:
> (Personally I hate all the fs interactions. There are too many ways to fail - think about owner, file mode, file systems)

Exactly, too many pieces for comfort for me.